examples/research_projects/jax-projects/README.md

```diff
@@ -383,13 +383,212 @@ official [flax example folder](https://github.com/huggingface/transformers/tree/
+
+      output = model(inputs, params=params)
+
+
+Let's look at an example to see how this works. First, we will write a simple two-layer MLP model.
+
+We will first write a flax module that will declare the layers and computation.
+
+```python
+import flax.linen as nn
+import jax.numpy as np
+
+class MLPMoudle(nn.Module):
+   config: MLPConfig
+   dtype: jnp.dtype = jnp.float32
+
+   def setup(self):
+      self.dense1 = nn.Dense(self.config.hidden_dim, dtype=self.dtype)
+      self.dense2 = nn.Desne(self.config.hidden_dim, dtype=self.dtype)
+   
+   def __call__(self, inputs):
+      hidden_states = self.dense1(inputs)
+      hidden_states = nn.relu(hidden_states)
+      hidden_states = self.dense2(hidden_states)
+      return hidden_states
+```
+
+Now let's define the `PreTrained` model class.
+
+```python
+from transformers.modeling_flax_utils import FlaxPreTrainedModel
+
+class FlaxMLPPreTrainedModel(FlaxPreTrainedModel):
+   config_class = MLPConfig
+   base_model_prefix = "model"
+   module_class: nn.Module = None
+
+   def __init__(self, config: BertConfig, input_shape: Tuple = (1, 8), seed: int = 0, dtype: jnp.dtype = jnp.float32, **kwargs):
+      # initialize the flax module
+      module = self.module_class(config=config, dtype=dtype, **kwargs)
+      super().__init__(config, module, input_shape=input_shape, seed=seed, dtype=dtype)
+   
+   def init_weights(self, rng, input_shape):
+      # init input tensors
+      inputs = jnp.zeros(input_shape, dtype="i4")
+      
+      params_rng, dropout_rng = jax.random.split(rng)
+      rngs = {"params": params_rng, "dropout": dropout_rng}
+      
+      params = self.module.init(rngs, inputs)["params"]
+      return params
+   
+   def __call__(self, inputs, params: dict = None):
+      params = {"params": params or self.params}
+      outputs = self.module.apply(params, jnp.array(inputs))
+      return outputs
+```
+
+
+Now we can define our model class as follows.
+
+```python
+class FlaxMLPModel(FlaxMLPPreTrainedModel):
+   module_class = FlaxMLPModule
+```
+
+Now the `FlaxMLPModel` will have a similar interface as that of other Transformers models.
+
+So the important point to remember is that the `model` is not an instance of `nn.Moudle`; it's an abstract class, like a container that holds a flax module, its parameters and provides convenient methods for initialization and forward pass. So our flax module is still stateless and immutable, but now we have a simple interface to work with. Feel free to take a look at the code to see how exactly this is implemented for ex. [`modeling_flax_bert.py`](https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_flax_bert.py)
```

Comment:
Also it would be nice to add one sentence like (to really hammer the point of stateless/statefull down):

```
So our flax module is still stateless and immutable, but now we have a simple interface to work with
```

=> The key take-away here is that an instance of `FlaxMLPModel` is very much stateful now since it holds all the model parameters, whereas the underlying Flax module `FlaxMLPModule` is still stateless. Now to make `FlaxMLPModel` fully compliant with JAX transformations, it is always possible to pass the parameters to `FlaxMLPModel` as well to make `FlaxMLPModel` stateless and easier to work with during training.